### PR TITLE
fix: [CDS-77413]: Add support for arrays as webhook json payload

### PR DESCRIPTION
**Pull request by vcalasansh**
## Describe your changes
Previously, we were only supporting webhook payload being a key-value map. This PR adds support for arrays as well.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51888)

### CODE REVIEW

**Comment by vcalasansh** [comment]
trigger messagecheck

**Comment by vcalasansh** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java` line 70 at r1](https://reviewable.io/reviews/harness/harness-core/51888#-NcVhovN0ofgEyEojRPu:-NcVhovN0ofgEyEojRPv:b-my8pgz) ([raw file](https://github.com/harness/harness-core/blob/111a7795e063df074fe742b6c85145516d8bec41/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java#L70)):*
> ```Java
>       } catch (IOException toHashMapEx) {
>         try {
>           jsonObject.put(PAYLOAD, JsonPipelineUtils.read(metadata.getTriggerJsonPayload(), ArrayList.class));
> ```

when can this be an array?

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java` line 70 at r1](https://reviewable.io/reviews/harness/harness-core/51888#-NcVhovN0ofgEyEojRPu:-NcVhsB36WeIf4Ll7Z9k:blq9ybl) ([raw file](https://github.com/harness/harness-core/blob/111a7795e063df074fe742b6c85145516d8bec41/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java#L70)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

when can this be an array?
</blockquote></details>

Shouldnt payload be always be a key value pair?

**Review by sahilhindwani-harness — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51888)* status: 2 file(s) to review. 0 of 2 approvals obtained.

**Review by rathodmeetsatish — DISMISSED** [review]

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java [hunk 1]
```diff
@@ -64,8 +65,12 @@ public Object bind() {
       // payload
       try {
         jsonObject.put(PAYLOAD, JsonPipelineUtils.read(metadata.getTriggerJsonPayload(), HashMap.class));
-      } catch (IOException e) {
-        throw new InvalidRequestException("Event payload could not be converted to a hashmap");
+      } catch (IOException toHashMapEx) {
+        try {
+          jsonObject.put(PAYLOAD, JsonPipelineUtils.read(metadata.getTriggerJsonPayload(), ArrayList.class));
```
Comment (sribalij) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java:70]: Can we code this to interface?
List.class?

**Review by sribalij — COMMENTED** [review]
Added a few comments

File: pipeline-service/modules/ng-triggers/src/test/java/io/harness/ngtriggers/expressions/TriggerFunctorTest.java [hunk 2]
```diff
@@ -211,7 +212,27 @@ public void testGetMetadataWebhook() {
...
     assertThatThrownBy(() -> finalExpressionEvaluator.renderExpression("<+trigger.event>"))
         .isInstanceOf(InvalidRequestException.class)
-        .hasMessage("Event payload could not be converted to a hashmap");
+        .hasMessage("Event payload could not be converted to a hashmap or arraylist");
+  }
+
+  @Test
+  @Owner(developers = VINICIUS)
+  @Category(UnitTests.class)
+  public void testCustomWebhookWhenPayloadIsArray() {
```
[the remark this answers is not in the document]
Comment (vcalasansh) [reply pipeline-service/modules/ng-triggers/src/test/java/io/harness/ngtriggers/expressions/TriggerFunctorTest.java:221]: Done. Thanks for the suggestion!

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java (hunk 1 above)
Comment (vcalasansh) [reply pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java]: Done

**Comment by vcalasansh** [comment]
trigger uts

**Review by sribalij — APPROVED** [review]

**Comment by sribalij** [comment]
trigger review

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51888)* status: 2 file(s) to review. 2. unresolved discussions.

**Comment by vcalasansh** [comment]
trigger review

**Pull request merged**